text/0000-reseeding-rand.md

```diff
@@ -0,0 +1,657 @@
+step (like `sample`), taking an `ExactSizeIterator` would reduce this
+to a single random number in total and allow truncating the traversal,
+andlastly taking a `RandomAccessIterator` would give the same indexing
+behaviour as a slice. The RFC author is unsure about the best path
+here, but notes that `Iterable` may make it possible to generalise
+this backwards-compatibly in future.
+
+The current `Rng` trait has two pieces of functionality that will be
+removed (or at least made `unstable`):
+
+- `rng.gen_weight_bool(n)`, this is trivially `rng.gen(0..n) == 0`.
+- `rng.gen_ascii_chars()`, this seems relatively specialised, and the
+  full range can be simulated with `rng.gen_iter('0'..'{')` (NB. `{`
+  is the char after `z`, necessary due to exclusivity of the RHS)
+  which unfortunately picks up some punctuation in the middle, but
+  most use cases will be perfectly well served by just
+  `rng.gen_iter('a'..'{')`. We could also provide extra
+  implementations, such as `struct PrintableAscii; impl
+  Random<PrintableAscii> for char`, allowing
+  `rng.gen_iter(PrintableAscii)`.
+
+#### Alternatives
+
+This trait doesn't need to exist, everything could be free functions
+in the `rand` module.
+
+There could be a `fn by_ref<'a>(&'a mut self) -> ByRef<'a, Self>`
+function, like occurs for iterators and in IO, to provide a real type
+backing up `impl<'a, T: ?Sized + Rng> Rng for &'a mut T {}`.
+
+## Free functions
+
+[free functions]: #free-functions
+
+Top level convenience functions are provided:
+
+```rust
+fn thread_rng() -> ThreadRng {
+    // fetch the cached RNG from TLS, or create it if it doesn't exist
+}
+
+fn random<T: Random<D>, D>(data: D) -> T {
+    thread_rng().gen(data)
+}
+```
+
+### Alternatives
+
+There is valid concern that the choice of RNGs described below leaves
+users looking at jargony names for RNGs, which no particularly obvious
+default choice; possibly leading to poor selection. Functions like:
+
+```rust
+fn secure_rng() -> OsRng { ... }
+fn balanced_rng() -> ChaChaRng { ... }
+fn weak_rng() -> XorShiftRng { ... }
+```
+
+could be provided to give some guidance. As a counterpoint, the
+`collections` library does not provide `ordered_map() -> BTreeMap`,
+`unordered_map() -> HashMap`.
+
+## Error handling
+
+[error handling]: #error-handling
+
+As it stands, there is no error handling at all, which is
+unfortunate. Most PRNGs do not have any error conditions (some define
+reaching the end of the sequence and repeating as an error), but the
+interface with the operating system may.
+
+Reconciling this is similar to IO error handling, where some readers
+and writers are guaranteed to not return errors (e.g. reading from
+memory). This consideration is postponed until the design in
+[RFC 576](https://github.com/rust-lang/rfcs/pull/576) is definite, but
+it would likely be along the lines of making each function return
+`Result<..., <Self as Reader>::Error>`. instead of just `...`. This
```

Comment:
I don't quite understand what you mean; presumably we can't impose `Err = Void` for all RNGs, because that would defeat the purpose of using `Result` at all.